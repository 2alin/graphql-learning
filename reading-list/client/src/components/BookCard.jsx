import React from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { getBooksQuery, removeBookMutation } from '../queries';

const RemoveBook = id => (
  <Mutation mutation={removeBookMutation} refetchQueries={[{ query: getBooksQuery }]}>
    {removeBook => (
      <button type="button" onClick={() => removeBook({ variables: { id } })} className="remove">
        X
      </button>
    )}
  </Mutation>
);

function BookCard({ book, handleSelect }) {
  return (
    <div className="card-container">
      <button type="button" onClick={handleSelect} className="card">
        {book.name}
      </button>
      {RemoveBook(book.id)}
    </div>
  );
}

BookCard.propTypes = {
  book: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }).isRequired,
  handleSelect: PropTypes.func.isRequired,
};

export default BookCard;
