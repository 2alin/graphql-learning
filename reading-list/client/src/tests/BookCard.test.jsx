import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from 'react-apollo/test-utils';
import BookCard from '../components/BookCard';
import { getBooksQuery, removeBookMutation } from '../queries';

const mocks = [
  {
    request: {
      query: getBooksQuery,
    },
    result: {
      data: { books: [] },
    },
  },
  {
    request: {
      query: removeBookMutation,
      variables: { id: '123' },
    },
    result: {
      data: {
        removeBook: { name: 'My Little Book', id: '123' },
      },
    },
  },
];

const book = {
  name: 'The Little Book',
  id: '123',
};
const handleSelect = id => id;

const component = renderer.create(
  <MockedProvider mocks={mocks} addTypename={false}>
    <BookCard {...{ book, handleSelect }} />
  </MockedProvider>,
);

it('component renders correctly', () => {
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

describe('describe `card`', () => {
  it('renders correctly', () => {
    const card = component.root.findByProps({ className: 'card' });
    expect(card).toBeTruthy();
  });

  it('renders book name', () => {
    const card = component.root.findByProps({ className: 'card' });
    expect(card.children).toContain(book.name);
  });

  it('onClick triggers an event with `id` as argument', () => {
    const card = component.root.findByProps({ className: 'card' });
    expect(card.props.onClick()).toBe(book.id);
  });
});

describe('describe `remove` button', () => {
  it('renders correctly', () => {
    const removeButton = component.root.findByProps({ className: 'remove' });
    expect(removeButton).toBeTruthy();
  });

  it('onClick resolves `removeBookMutation`', async () => {
    const removeButton = component.root.findByProps({ className: 'remove' });
    const res = removeButton.props.onClick();
    await expect(res).resolves.toEqual({
      data: {
        removeBook: {
          name: 'My Little Book',
          id: '123',
        },
      },
    });
  });
});
