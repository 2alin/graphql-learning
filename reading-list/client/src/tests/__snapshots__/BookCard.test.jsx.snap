// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders correctly 1`] = `
<div
  className="card-container"
>
  <button
    className="card"
    onClick={[Function]}
    type="button"
  >
    The Little Book
  </button>
  <button
    className="remove"
    onClick={[Function]}
    type="button"
  >
    X
  </button>
</div>
`;
